pkg/sdkserver/sdkserver.go

```diff
@@ -762,20 +789,190 @@ func (s *SDKServer) GetCounter(ctx context.Context, in *alpha.GetCounterRequest)
 		return nil, errors.Errorf("%s not enabled", runtime.FeatureCountsAndLists)
 	}
-	// TODO(#2716): Implement me
-	return nil, errors.Errorf("Unimplemented -- GetCounter coming soon")
+
+	s.logger.WithField("name", in.Name).Debug("Getting Counter")
+
+	gs, err := s.gameServer()
+	if err != nil {
+		return nil, err
+	}
+
+	s.gsUpdateMutex.RLock()
+	defer s.gsUpdateMutex.RUnlock()
+
+	counter, ok := gs.Status.Counters[in.Name]
+	if !ok {
+		return nil, errors.Errorf("NOT_FOUND. %s Counter not found", in.Name)
+	}
+	s.logger.WithField("Get Counter", counter).Debugf("Got Counter %s", in.Name)
+	protoCounter := alpha.Counter{Name: in.Name, Count: counter.Count, Capacity: counter.Capacity}
+	// If there are batched changes that have not yet been applied, apply them to the Counter.
+	// This does NOT validate batched the changes.
+	if counterUpdate, ok := s.gsCounterUpdates[in.Name]; ok {
+		if counterUpdate.capacitySet != nil {
+			protoCounter.Capacity = *counterUpdate.capacitySet
+		}
+		if counterUpdate.countSet != nil {
+			protoCounter.Count = *counterUpdate.countSet
+		}
+		protoCounter.Count += counterUpdate.diff
+		if protoCounter.Count < 0 {
+			protoCounter.Count = 0
+			s.logger.Debug("truncating Count in Get Counter request to 0")
+		}
+		if protoCounter.Count > protoCounter.Capacity {
+			protoCounter.Count = protoCounter.Capacity
+			s.logger.Debug("truncating Count in Get Counter request to Capacity")
+		}
+		s.logger.WithField("Get Counter", counter).Debugf("Applied Batched Counter Updates %v", counterUpdate)
+	}
+
+	return &protoCounter, nil
 }
 
-// UpdateCounter returns the updated Counter. Returns NOT_FOUND if the Counter does not exist.
-// Returns INVALID_ARGUMENT if the FieldMask paths are invalid.
+// UpdateCounter collapses all UpdateCounterRequests for a given Counter into a single request.
+// Returns NOT_FOUND if the Counter does not exist (name cannot be updated).
+// Returns OUT_OF_RANGE if the Count is out of range [0,Capacity].
 // [Stage:Alpha]
 // [FeatureFlag:CountsAndLists]
 func (s *SDKServer) UpdateCounter(ctx context.Context, in *alpha.UpdateCounterRequest) (*alpha.Counter, error) {
 	if !runtime.FeatureEnabled(runtime.FeatureCountsAndLists) {
 		return nil, errors.Errorf("%s not enabled", runtime.FeatureCountsAndLists)
 	}
-	// TODO(#2716): Implement Me
-	return nil, errors.Errorf("Unimplemented -- UpdateCounter coming soon")
+
+	if in.CounterUpdateRequest == nil {
+		return nil, errors.Errorf("INVALID_ARGUMENT. CounterUpdateRequest: %v cannot be nil", in.CounterUpdateRequest)
+	}
+
+	s.logger.WithField("name", in.CounterUpdateRequest.Name).Debug("Update Counter Request")
+	s.gsUpdateMutex.Lock()
+	defer s.gsUpdateMutex.Unlock()
+
+	// Check if we already have a batch request started for this Counter. If not, add new request to
+	// the gsCounterUpdates map.
+	name := in.CounterUpdateRequest.Name
+	batchCounter, batchOk := s.gsCounterUpdates[name]
+
+	if !batchOk {
```

Comment:
```suggestion
```

I'm thinking we should get rid of this outer if statement and always get the GameServer, and update the underlying `counter` values. That way _if_ there is a value that comes through from an allocation while a batch is yet to be processed, it has a chance to get captured, whereas this way it will get ignored, and if it's the same as always then it's a no-op (and it's an in-memory cache, so it doesn't matter).

I think that will remove a potential race condition.

Reply:
Sure, we can do that.